### ISSUE TITLE
InvalidReturnType emitted in case of implemented method with generics traversable return type

### ISSUE THREAD

**Issue by andrew-demb** [vimeo/psalm#7626]
https://psalm.dev/r/1d31e9718a

If I drop generic template from class - no error is triggered
https://psalm.dev/r/1a62cfd31b

**Comment by AndrolGenhald** [comment]
This is correct, you're always returning a `RedisStorageEntity` no matter what type `T` is. Do you have a slightly longer snippet that runs on 3v4l.org? With the example you provide there's no use for the template on `RedisStorage` anyway.

**Comment by orklah** [comment]
I agree, this seems correct. If you declare a T and you use it in the implements, then `getByIds` MUST return T, not any other RedisStorageEntity.

For example, if I create two classes A and B, both extending RedisStorageEntity, the fact that you declared a T template would mean I could have a `RedisStorage<A>`. When I call `getByIds` on `RedisStorage<A>`, it must return a A, not a B

**Comment by andrew-demb** [comment]
Thank you for clarifying.

Real code was with serializer objects (but without corresponding templating as I refactor step by step code).

For history:
"Real" code from our codebase: https://psalm.dev/r/4142ed535b
Valid code from Psalm PoV: https://psalm.dev/r/2bc1c62a85
